File: bindings/python/python/auths/conformance.py
from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass
from inspect import isawaitable
from typing import (
    Awaitable,
    Callable,
    Literal,
    Mapping,
    Protocol,
    Sequence,
    Tuple,
    Union,
)

from ._mechanism_conformance import CONFORMANCE_CATALOG
from .custody import ProviderOperationError, Signer, SigningRequest
from .integrations import development
from .profiles import mcp
from .profiles.mcp import McpClosedProvider, McpHandlerOutcome, McpToolContext
from .workflow import ReviewField


@dataclass(frozen=True)
class ConformanceMetadata:
    implementation: str
    version: str
    runtime: str = "python"
    capabilities: Tuple[str, ...] = ()


@dataclass(frozen=True)
class ConformanceCaseResult:
    id: str
    classification: Literal["deterministic"]
    passed: bool


@dataclass(frozen=True)
class ConformanceReport:
    schema: Literal["auths.conformance-report/1"]
    suite: str
    suite_version: int
    semantic_subject: Literal["auths.mechanism-profile-conformance/1"]
    implementation: str
    implementation_version: str
    runtime: str
    capabilities: Tuple[str, ...]
    results: Tuple[ConformanceCaseResult, ...]
    passed: bool
    claim: Literal["test-results-only-not-security-certification"]


@dataclass(frozen=True)
class AtomicReservationRecord:
    key: str
    commitment: bytes
    value: bytes


class AtomicReservationStoreCandidate(Protocol):
    async def reserve(
        self, record: AtomicReservationRecord
    ) -> Literal["acquired", "exact-replay", "conflict"]: ...

    async def aclose(self) -> None: ...

    async def reopen(self) -> AtomicReservationStoreCandidate: ...


class ByteTransportCandidate(Protocol):
    async def exchange(
        self,
        packet: bytes,
        *,
        maximum_bytes: int,
        cancellation: asyncio.Event,
    ) -> bytes: ...

    async def aclose(self) -> None: ...


FactoryResult = Union[object, Awaitable[object]]
AtomicStoreFactory = Callable[[], FactoryResult]
ByteTransportFactory = Callable[[Callable[[bytes], Awaitable[bytes]]], FactoryResult]
McpProviderFactory = Callable[..., McpClosedProvider]


async def certify_signer(
    factory: Callable[[], FactoryResult], metadata: ConformanceMetadata
) -> ConformanceReport:
    outcomes = []
    for case in (
        "transaction-binding",
        "principal-binding",
        "descriptor-binding",
        "request-binding",
    ):
        outcomes.append((f"signer/{case}", await _signer_binding_case(factory, case)))
    outcomes.append(("signer/expiry", await _signer_rejection_case(factory, "expiry")))
    outcomes.append(
        ("signer/duplicate", await _signer_rejection_case(factory, "duplicate"))
    )
    outcomes.append(
        ("signer/cancellation", await _signer_rejection_case(factory, "cancellation"))
    )
    outcomes.append(
        ("signer/disposal", await _signer_rejection_case(factory, "disposal"))
    )
    return _report("signer-custody/1", metadata, outcomes)


async def certify_atomic_store(
    factory: AtomicStoreFactory, metadata: ConformanceMetadata
) -> ConformanceReport:
    record = _reservation("case", 1, b"\x03")
    outcomes = [
        (
            "atomic-store/acquire",
            await _atomic_case(
                factory, lambda store: _equals_reservation(store, record, "acquired")
            ),
        ),
        (
            "atomic-store/exact-replay",
            await _atomic_case(factory, lambda store: _exact_replay(store, record)),
        ),
        (
            "atomic-store/conflict",
            await _atomic_case(factory, lambda store: _conflict(store, record)),
        ),
        (
            "atomic-store/concurrent-single-winner",
            await _atomic_case(factory, lambda store: _concurrent(store, record)),
        ),
        (
            "atomic-store/bounded-record",
            await _atomic_case(factory, _bounded_record),
        ),
    ]
    first = await _resolve(factory())
    second = await _resolve(factory())
    try:
        isolated = (
            await first.reserve(record) == "acquired"
            and await second.reserve(record) == "acquired"
        )
    except Exception:
        isolated = False
    finally:
        await _close(first)
        await _close(second)
    outcomes.append(("atomic-store/isolated-instances", isolated))
    outcomes.append(
        (
            "atomic-store/reopen-durability-claim",
            await _durability_case(factory, record, metadata),
        )
    )
    return _report("atomic-reservation-store/1", metadata, outcomes)


async def certify_byte_transport(
    factory: ByteTransportFactory, metadata: ConformanceMetadata
) -> ConformanceReport:
    async def exact(packet: bytes) -> bytes:
        return packet

    async def oversized(_: bytes) -> bytes:
        return b"x" * 17

    outcomes = [
        (
            "byte-transport/exact-bytes",
            await _transport_case(
                factory,
                exact,
                lambda transport: _exchange_equals(transport, b"abc", 16, b"abc"),
            ),
        ),
        (
            "byte-transport/bounded-input",
            await _transport_case(
                factory,
                exact,
                lambda transport: _exchange_rejected(transport, b"x" * 17, 16),
            ),
        ),
        (
            "byte-transport/bounded-output",
            await _transport_case(
                factory,
                oversized,
                lambda transport: _exchange_rejected(transport, b"x", 16),
            ),
        ),
        (
            "byte-transport/cancellation",
            await _transport_case(factory, exact, _cancelled_exchange),
        ),
        (
            "byte-transport/disposal",
            await _transport_case(factory, exact, _disposed_exchange, close=False),
        ),
    ]
    return _report("bounded-byte-transport/1", metadata, outcomes)


async def certify_mcp_provider(
    factory: McpProviderFactory, metadata: ConformanceMetadata
) -> ConformanceReport:
    outcomes = []
    calls = 0
    request_bound = False

    async def publish(
        arguments: Mapping[str, object], context: McpToolContext
    ) -> object:
        nonlocal calls, request_bound
        calls += 1
        request_bound = (
            arguments.get("report") == "weekly"
            and context.service == "development"
            and context.tool == "publish_report"
        )
        return {"ok": True}

    auths = await development.create_auths(
        authority=mcp.allow_tools(["publish_report"])
    )
    try:
        action = mcp.call_tool(name="publish_report", arguments={"report": "weekly"})
        exact = factory(tools={"publish_report": publish})
        completed = await auths.execute(
            action=action, provider=exact, request_id="conformance-exact"
        )
        outcomes.append(
            (
                "mcp/exact-call",
                completed.kind == "completed" and calls == 1 and request_bound,
            )
        )
        denied = await auths.execute(
            action=mcp.call_tool(name="delete_report", arguments={}),
            provider=exact,
            request_id="conformance-denied",
        )
        outcomes.append(
            ("mcp/deny-before-entry", denied.kind == "denied" and calls == 1)
        )
        concurrent = await asyncio.gather(
            auths.execute(
                action=action,
                provider=exact,
                request_id="conformance-concurrent",
            ),
            auths.execute(
                action=action,
                provider=exact,
                request_id="conformance-concurrent",
            ),
        )
        outcomes.append(
            (
                "mcp/concurrent-single-entry",
                sorted(value.kind for value in concurrent)
                == ["completed", "exact-replay"]
                and calls == 2,
            )
        )
    except Exception:
        _ensure_outcomes(
            outcomes,
            "mcp/exact-call",
            "mcp/deny-before-entry",
            "mcp/concurrent-single-entry",
        )
    finally:
        await auths.aclose()

    ambiguous_calls = 0

    async def ambiguous(
        arguments: Mapping[str, object], context: McpToolContext
    ) -> object:
        nonlocal ambiguous_calls
        ambiguous_calls += 1
        return McpHandlerOutcome("possible", cause="unknown")

    async def forbidden(
        arguments: Mapping[str, object], context: McpToolContext
    ) -> object:
        nonlocal ambiguous_calls
        ambiguous_calls += 100
        return None

    async def reconcile(execution_id: str, service: str) -> McpHandlerOutcome[object]:
        return McpHandlerOutcome("applied", {"ok": True})

    recovery = await development.create_auths(
        authority=mcp.allow_tools(["publish_report"])
    )
    try:
        pending = await recovery.execute(
            action=mcp.call_tool(name="publish_report", arguments={}),
            provider=factory(tools={"publish_report": ambiguous}),
            request_id="conformance-recovery",
        )
        outcomes.append(
            (
                "mcp/ambiguous-no-blind-retry",
                pending.kind == "recoverable" and ambiguous_calls == 1,
            )
        )
        resumed = await recovery.resume(
            reference=pending.reference,
            provider=factory(tools={"publish_report": forbidden}, reconcile=reconcile),
        )
        outcomes.append(
            (
                "mcp/reconcile-without-reentry",
                resumed.kind == "completed" and ambiguous_calls == 1,
            )
        )
    except Exception:
        _ensure_outcomes(
            outcomes,
            "mcp/ambiguous-no-blind-retry",
            "mcp/reconcile-without-reentry",
        )
    finally:
        await recovery.aclose()

    service_auths = await development.create_auths(
        authority=mcp.allow_tools(["publish_report"])
    )
    try:
        await service_auths.execute(
            action=mcp.call_tool(name="publish_report", arguments={}),
            provider=factory(service="other", tools={"publish_report": publish}),
        )
        outcomes.append(("mcp/service-binding", False))
    except Exception:
        outcomes.append(("mcp/service-binding", True))
    finally:
        await service_auths.aclose()

    async def oversized(
        arguments: Mapping[str, object], context: McpToolContext
    ) -> object:
        return "x" * 1_048_577

    bounded_auths = await development.create_auths(
        authority=mcp.allow_tools(["publish_report"])
    )
    try:
        result = await bounded_auths.execute(
            action=mcp.call_tool(name="publish_report", arguments={}),
            provider=factory(tools={"publish_report": oversized}),
        )
        outcomes.append(("mcp/bounded-output", result.kind != "completed"))
    except Exception:
        outcomes.append(("mcp/bounded-output", True))
    finally:
        await bounded_auths.aclose()
    return _report("auths.mcp/1/provider/1", metadata, outcomes)


async def _signer_binding_case(factory: Callable[[], FactoryResult], case: str) -> bool:
    signer = await _resolve(factory())
    try:
        request = await _signing_request(signer, case)
        response = await signer.sign(request)
        if case == "transaction-binding":
            return response.transaction_digest == request.transaction_digest
        if case == "principal-binding":
            return (
                response.principal.principal.value == request.principal.principal.value
            )
        if case == "descriptor-binding":
            return response.principal.matches(request.principal)
        return response.request_id == request.request_id
    except Exception:
        return False
    finally:
        await _close(signer)


async def _signer_rejection_case(
    factory: Callable[[], FactoryResult], case: str
) -> bool:
    signer = await _resolve(factory())
    try:
        request = await _signing_request(signer, case)
        if case == "expiry":
            request = SigningRequest(
                request.request_id,
                request.object_kind,
                request.object_id,
                request.principal,
                request.transaction_digest,
                request.signing_preimage,
                int(time.time()) - 1,
                request.display,
            )
            return await _provider_rejected(
                signer.sign(request), ("rejected", "timeout")
            )
        if case == "duplicate":
            await signer.sign(request)
            return await _provider_rejected(signer.sign(request), ("rejected",))
        if case == "cancellation":
            task = asyncio.create_task(signer.sign(request))
            task.cancel()
            try:
                await task
            except asyncio.CancelledError:
                return True
            return False
        await signer.aclose()
        return await _provider_rejected(
            signer.sign(request), ("cancelled", "unsupported")
        )
    except Exception:
        return False
    finally:
        await _close(signer)


async def _signing_request(signer: Signer, suffix: str) -> SigningRequest:
    principal = await signer.public_identity()
    return SigningRequest(
        f"conformance.{suffix}",
        "action",
        b"\x01" * 32,
        principal,
        b"\x02" * 32,
        b"\x03\x04\x05",
        int(time.time()) + 60,
        (ReviewField("Conformance", suffix),),
    )


async def _provider_rejected(
    awaitable: Awaitable[object], expected: Sequence[str]
) -> bool:
    try:
        await awaitable
        return False
    except ProviderOperationError as error:
        return error.kind in expected
    except Exception:
        return False


async def _atomic_case(
    factory: AtomicStoreFactory,
    exercise: Callable[[AtomicReservationStoreCandidate], Awaitable[bool]],
) -> bool:
    store = await _resolve(factory())
    try:
        return await exercise(store)
    except Exception:
        return False
    finally:
        await _close(store)


async def _equals_reservation(
    store: AtomicReservationStoreCandidate,
    record: AtomicReservationRecord,
    expected: str,
) -> bool:
    return await store.reserve(record) == expected


async def _exact_replay(
    store: AtomicReservationStoreCandidate, record: AtomicReservationRecord
) -> bool:
    return (
        await store.reserve(record) == "acquired"
        and await store.reserve(record) == "exact-replay"
    )


async def _conflict(
    store: AtomicReservationStoreCandidate, record: AtomicReservationRecord
) -> bool:
    return (
        await store.reserve(record) == "acquired"
        and await store.reserve(_reservation("case", 2, b"\x04")) == "conflict"
    )


async def _concurrent(
    store: AtomicReservationStoreCandidate, record: AtomicReservationRecord
) -> bool:
    values = await asyncio.gather(*(store.reserve(record) for _ in range(8)))
    return values.count("acquired") == 1 and values.count("exact-replay") == 7


async def _bounded_record(store: AtomicReservationStoreCandidate) -> bool:
    try:
        await store.reserve(_reservation("oversized", 1, b"x" * 262_145))
        return False
    except (Exception, asyncio.CancelledError):
        return True


async def _durability_case(
    factory: AtomicStoreFactory,
    record: AtomicReservationRecord,
    metadata: ConformanceMetadata,
) -> bool:
    if "durable-reopen" not in metadata.capabilities:
        return True
    first = await _resolve(factory())
    second = None
    try:
        if await first.reserve(record) != "acquired":
            return False
        reopen = getattr(first, "reopen", None)
        if not callable(reopen):
            return False
        second = await _resolve(reopen())
        return await second.reserve(record) == "exact-replay"
    except Exception:
        return False
    finally:
        if second is not None:
            await _close(second)
        await _close(first)


async def _transport_case(
    factory: ByteTransportFactory,
    deliver: Callable[[bytes], Awaitable[bytes]],
    exercise: Callable[[ByteTransportCandidate], Awaitable[bool]],
    *,
    close: bool = True,
) -> bool:
    transport = await _resolve(factory(deliver))
    try:
        return await exercise(transport)
    except Exception:
        return False
    finally:
        if close:
            await _close(transport)


async def _exchange_equals(
    transport: ByteTransportCandidate,
    packet: bytes,
    maximum: int,
    expected: bytes,
) -> bool:
    return (
        await transport.exchange(
            packet, maximum_bytes=maximum, cancellation=asyncio.Event()
        )
        == expected
    )


async def _exchange_rejected(
    transport: ByteTransportCandidate, packet: bytes, maximum: int
) -> bool:
    try:
        await transport.exchange(
            packet, maximum_bytes=maximum, cancellation=asyncio.Event()
        )
        return False
    except (Exception, asyncio.CancelledError):
        return True


async def _cancelled_exchange(transport: ByteTransportCandidate) -> bool:
    cancellation = asyncio.Event()
    cancellation.set()
    try:
        await transport.exchange(b"x", maximum_bytes=16, cancellation=cancellation)
        return False
    except (asyncio.CancelledError, Exception):
        return True


async def _disposed_exchange(transport: ByteTransportCandidate) -> bool:
    await transport.aclose()
    return await _exchange_rejected(transport, b"x", 16)


def _report(
    suite: str,
    metadata: ConformanceMetadata,
    outcomes: Sequence[Tuple[str, bool]],
) -> ConformanceReport:
    expected = next(
        (
            candidate
            for candidate in CONFORMANCE_CATALOG["suites"]
            if candidate["id"] == suite
        ),
        None,
    )
    if expected is None:
        raise ValueError("unknown Auths conformance suite")
    supplied = dict(outcomes)
    results = tuple(
        ConformanceCaseResult(
            candidate["id"],
            candidate["classification"],
            supplied.get(candidate["id"]) is True,
        )
        for candidate in expected["cases"]
    )
    capabilities = tuple(
        _bounded(value, "capability") for value in metadata.capabilities
    )
    return ConformanceReport(
        "auths.conformance-report/1",
        suite,
        1,
        "auths.mechanism-profile-conformance/1",
        _bounded(metadata.implementation, "implementation"),
        _bounded(metadata.version, "implementation version"),
        _bounded(metadata.runtime, "runtime"),
        capabilities,
        results,
        all(result.passed for result in results),
        "test-results-only-not-security-certification",
    )


def _reservation(key: str, byte: int, value: bytes) -> AtomicReservationRecord:
    return AtomicReservationRecord(key, bytes([byte]) * 32, bytes(value))


async def _resolve(value: FactoryResult):
    return await value if isawaitable(value) else value


async def _close(value: object) -> None:
    close = getattr(value, "aclose", None)
    if callable(close):
        await close()


def _ensure_outcomes(outcomes: list, *case_ids: str) -> None:
    present = {identifier for identifier, _ in outcomes}
    outcomes.extend(
        (identifier, False) for identifier in case_ids if identifier not in present
    )


def _bounded(value: str, name: str) -> str:
    if (
        type(value) is not str
        or not value
        or len(value) > 128
        or "\n" in value
        or "\r" in value
    ):
        raise ValueError(f"invalid conformance {name}")
    return value


__all__ = [
    "AtomicReservationRecord",
    "AtomicReservationStoreCandidate",
    "ByteTransportCandidate",
    "CONFORMANCE_CATALOG",
    "ConformanceCaseResult",
    "ConformanceMetadata",
    "ConformanceReport",
    "certify_atomic_store",
    "certify_byte_transport",
    "certify_mcp_provider",
    "certify_signer",
]

File: bindings/typescript/src/testkit/conformance.ts
import type { Signer } from "../custody.js";
import { development } from "../integrations.js";
import {
  mcp,
  type McpClosedProvider,
  type McpHandlerOutcome,
  type McpToolContext,
} from "../mcp.js";
import { CONFORMANCE_CATALOG } from "../generated/mechanism-conformance.js";
import { custodyConformance } from "./custody-conformance.js";

export interface ConformanceMetadata {
  readonly implementation: string;
  readonly version: string;
  readonly runtime?: string;
  readonly capabilities?: readonly string[];
}

export interface ConformanceCaseResult {
  readonly id: string;
  readonly classification: "deterministic";
  readonly passed: boolean;
}

export interface ConformanceReport {
  readonly schema: "auths.conformance-report/1";
  readonly suite: string;
  readonly suiteVersion: 1;
  readonly semanticSubject: "auths.mechanism-profile-conformance/1";
  readonly implementation: string;
  readonly implementationVersion: string;
  readonly runtime: string;
  readonly capabilities: readonly string[];
  readonly results: readonly ConformanceCaseResult[];
  readonly passed: boolean;
  readonly claim: "test-results-only-not-security-certification";
}

export interface AtomicReservationRecord {
  readonly key: string;
  readonly commitment: Uint8Array;
  readonly value: Uint8Array;
}

export interface AtomicReservationStoreCandidate {
  reserve(record: AtomicReservationRecord): Promise<"acquired" | "exact-replay" | "conflict">;
  reopen?(): AtomicReservationStoreCandidate | Promise<AtomicReservationStoreCandidate>;
  close?(): void | Promise<void>;
}

export interface ByteTransportCandidate {
  exchange(
    packet: Uint8Array,
    options: Readonly<{ maximumBytes: number; signal: AbortSignal }>,
  ): Promise<Uint8Array>;
  close?(): void | Promise<void>;
}

export type ByteTransportFactory = (
  deliver: (packet: Uint8Array) => Uint8Array | Promise<Uint8Array>,
) => ByteTransportCandidate | Promise<ByteTransportCandidate>;

export type McpProviderFactory = (options: Readonly<{
  service?: string;
  tools: Readonly<Record<string, (
    argumentsValue: Readonly<Record<string, unknown>>,
    context: McpToolContext,
  ) => unknown | Promise<unknown>>>;
  reconcile?: (executionId: string, service: string) => McpHandlerOutcome<unknown> | Promise<McpHandlerOutcome<unknown>>;
}>) => McpClosedProvider;

export async function certifySigner(
  factory: () => Signer | Promise<Signer>,
  metadata: ConformanceMetadata,
): Promise<ConformanceReport> {
  const observed = await custodyConformance({ create: async () => factory() });
  return report(
    "signer-custody/1",
    metadata,
    observed.results.map((result) => [`signer/${result.name}`, result.passed] as const),
  );
}

export async function certifyAtomicStore(
  factory: () => AtomicReservationStoreCandidate | Promise<AtomicReservationStoreCandidate>,
  metadata: ConformanceMetadata,
): Promise<ConformanceReport> {
  const record = reservation("case", 1, new Uint8Array([3]));
  const outcomes: Array<readonly [string, boolean]> = [];
  outcomes.push(["atomic-store/acquire", await atomicCase(factory, async (store) =>
    await store.reserve(record) === "acquired")]);
  outcomes.push(["atomic-store/exact-replay", await atomicCase(factory, async (store) =>
    await store.reserve(record) === "acquired" && await store.reserve(record) === "exact-replay")]);
  outcomes.push(["atomic-store/conflict", await atomicCase(factory, async (store) =>
    await store.reserve(record) === "acquired"
      && await store.reserve(reservation("case", 2, new Uint8Array([4]))) === "conflict")]);
  outcomes.push(["atomic-store/concurrent-single-winner", await atomicCase(factory, async (store) => {
    const values = await Promise.all(Array.from({ length: 8 }, () => store.reserve(record)));
    return values.filter((value) => value === "acquired").length === 1
      && values.filter((value) => value === "exact-replay").length === 7;
  })]);
  outcomes.push(["atomic-store/bounded-record", await atomicCase(factory, async (store) => {
    try {
      await store.reserve(reservation("oversized", 1, new Uint8Array(262_145)));
      return false;
    } catch {
      return true;
    }
  })]);
  const first = await factory();
  const second = await factory();
  try {
  outcomes.push(["atomic-store/isolated-instances",
      await first.reserve(record) === "acquired" && await second.reserve(record) === "acquired"]);
  } catch {
    outcomes.push(["atomic-store/isolated-instances", false]);
  } finally {
    await first.close?.();
    await second.close?.();
  }
  outcomes.push(["atomic-store/reopen-durability-claim", await durabilityCase(factory, record, metadata)]);
  return report("atomic-reservation-store/1", metadata, outcomes);
}

export async function certifyByteTransport(
  factory: ByteTransportFactory,
  metadata: ConformanceMetadata,
): Promise<ConformanceReport> {
  const outcomes: Array<readonly [string, boolean]> = [];
  outcomes.push(["byte-transport/exact-bytes", await transportCase(factory, (packet) => packet, async (transport) => {
    const value = new Uint8Array([1, 2, 3]);
    return equalBytes(await transport.exchange(value, transportOptions(16)), value);
  })]);
  outcomes.push(["byte-transport/bounded-input", await transportCase(factory, (packet) => packet, async (transport) => {
    try {
      await transport.exchange(new Uint8Array(17), transportOptions(16));
      return false;
    } catch {
      return true;
    }
  })]);
  outcomes.push(["byte-transport/bounded-output", await transportCase(factory, () => new Uint8Array(17), async (transport) => {
    try {
      await transport.exchange(new Uint8Array([1]), transportOptions(16));
      return false;
    } catch {
      return true;
    }
  })]);
  outcomes.push(["byte-transport/cancellation", await transportCase(factory, (packet) => packet, async (transport) => {
    const controller = new AbortController();
    controller.abort();
    try {
      await transport.exchange(new Uint8Array([1]), { maximumBytes: 16, signal: controller.signal });
      return false;
    } catch {
      return true;
    }
  })]);
  outcomes.push(["byte-transport/disposal", await transportCase(factory, (packet) => packet, async (transport) => {
    if (transport.close === undefined) return false;
    await transport.close();
    try {
      await transport.exchange(new Uint8Array([1]), transportOptions(16));
      return false;
    } catch {
      return true;
    }
  }, false)]);
  return report("bounded-byte-transport/1", metadata, outcomes);
}

export async function certifyMcpProvider(
  factory: McpProviderFactory,
  metadata: ConformanceMetadata,
): Promise<ConformanceReport> {
  const outcomes: Array<readonly [string, boolean]> = [];
  let calls = 0;
  let requestBound = false;
  const exact = factory({ tools: { async publish_report(argumentsValue, context) {
    calls += 1;
    requestBound = argumentsValue.report === "weekly"
      && context.service === "development"
      && context.tool === "publish_report";
    return { ok: true };
  } } });
  const auths = await development.createAuths({ authority: mcp.allowTools(["publish_report"]) });
  try {
    const action = mcp.callTool({ name: "publish_report", arguments: { report: "weekly" } });
    const completed = await auths.execute({ action, provider: exact, requestId: "conformance-exact" });
    outcomes.push(["mcp/exact-call", completed.kind === "completed" && calls === 1 && requestBound]);
    const denied = await auths.execute({
      action: mcp.callTool({ name: "delete_report", arguments: {} }),
      provider: exact,
      requestId: "conformance-denied",
    });
    outcomes.push(["mcp/deny-before-entry", denied.kind === "denied" && calls === 1]);
    const concurrent = await Promise.all([
      auths.execute({ action, provider: exact, requestId: "conformance-concurrent" }),
      auths.execute({ action, provider: exact, requestId: "conformance-concurrent" }),
    ]);
    outcomes.push(["mcp/concurrent-single-entry",
      concurrent.map((value) => value.kind).sort().join(",") === "completed,exact-replay" && calls === 2]);
  } catch {
    outcomes.push(["mcp/exact-call", false], ["mcp/deny-before-entry", false], ["mcp/concurrent-single-entry", false]);
  } finally {
    await auths.close();
  }

  let ambiguousCalls = 0;
  const recoveryAuths = await development.createAuths({ authority: mcp.allowTools(["publish_report"]) });
  try {
    const pending = await recoveryAuths.execute({
      action: mcp.callTool({ name: "publish_report", arguments: {} }),
      provider: factory({ tools: { async publish_report() {
        ambiguousCalls += 1;
        return { effect: "possible", cause: "unknown" };
      } } }),
      requestId: "conformance-recovery",
    });
    outcomes.push(["mcp/ambiguous-no-blind-retry", pending.kind === "recoverable" && ambiguousCalls === 1]);
    const resumed = pending.kind === "recoverable" && pending.reference !== undefined
      ? await recoveryAuths.resume({
        reference: pending.reference,
        provider: factory({
          tools: { async publish_report() { ambiguousCalls += 100; } },
          async reconcile() { return { effect: "applied", result: { ok: true } }; },
        }),
      })
      : undefined;
    outcomes.push(["mcp/reconcile-without-reentry", resumed?.kind === "completed" && ambiguousCalls === 1]);
  } catch {
    outcomes.push(["mcp/ambiguous-no-blind-retry", false], ["mcp/reconcile-without-reentry", false]);
  } finally {
    await recoveryAuths.close();
  }

  const serviceAuths = await development.createAuths({ authority: mcp.allowTools(["publish_report"]) });
  try {
    await serviceAuths.execute({
      action: mcp.callTool({ name: "publish_report", arguments: {} }),
      provider: factory({ service: "other", tools: { async publish_report() {} } }),
    });
    outcomes.push(["mcp/service-binding", false]);
  } catch {
    outcomes.push(["mcp/service-binding", true]);
  } finally {
    await serviceAuths.close();
  }

  const boundedAuths = await development.createAuths({ authority: mcp.allowTools(["publish_report"]) });
  try {
    const bounded = await boundedAuths.execute({
      action: mcp.callTool({ name: "publish_report", arguments: {} }),
      provider: factory({ tools: { async publish_report() { return "x".repeat(1_048_577); } } }),
    });
    outcomes.push(["mcp/bounded-output", bounded.kind !== "completed"]);
  } catch {
    outcomes.push(["mcp/bounded-output", true]);
  } finally {
    await boundedAuths.close();
  }
  return report("auths.mcp/1/provider/1", metadata, outcomes);
}

export { CONFORMANCE_CATALOG };

async function atomicCase(
  factory: () => AtomicReservationStoreCandidate | Promise<AtomicReservationStoreCandidate>,
  exercise: (store: AtomicReservationStoreCandidate) => Promise<boolean>,
): Promise<boolean> {
  const store = await factory();
  try {
    return await exercise(store);
  } catch {
    return false;
  } finally {
    await store.close?.();
  }
}

async function durabilityCase(
  factory: () => AtomicReservationStoreCandidate | Promise<AtomicReservationStoreCandidate>,
  record: AtomicReservationRecord,
  metadata: ConformanceMetadata,
): Promise<boolean> {
  if (!(metadata.capabilities ?? []).includes("durable-reopen")) return true;
  const first = await factory();
  let second: AtomicReservationStoreCandidate | undefined;
  try {
    if (await first.reserve(record) !== "acquired" || first.reopen === undefined) return false;
    second = await first.reopen();
    return await second.reserve(record) === "exact-replay";
  } catch {
    return false;
  } finally {
    await second?.close?.();
    await first.close?.();
  }
}

async function transportCase(
  factory: ByteTransportFactory,
  deliver: (packet: Uint8Array) => Uint8Array | Promise<Uint8Array>,
  exercise: (transport: ByteTransportCandidate) => Promise<boolean>,
  close = true,
): Promise<boolean> {
  const transport = await factory(deliver);
  try {
    return await exercise(transport);
  } catch {
    return false;
  } finally {
    if (close) await transport.close?.();
  }
}

function report(
  suite: string,
  metadata: ConformanceMetadata,
  outcomes: readonly (readonly [string, boolean])[],
): ConformanceReport {
  const expected = CONFORMANCE_CATALOG.suites.find((candidate) => candidate.id === suite);
  if (expected === undefined) throw new TypeError("unknown Auths conformance suite");
  const supplied = new Map(outcomes);
  const results = expected.cases.map((candidate) => Object.freeze({
    id: candidate.id,
    classification: candidate.classification,
    passed: supplied.get(candidate.id) === true,
  }));
  const implementation = bounded(metadata.implementation, "implementation");
  const version = bounded(metadata.version, "implementation version");
  const runtime = bounded(metadata.runtime ?? "javascript", "runtime");
  const capabilities = Object.freeze((metadata.capabilities ?? []).map((value) => bounded(value, "capability")));
  return Object.freeze({
    schema: "auths.conformance-report/1",
    suite,
    suiteVersion: 1,
    semanticSubject: "auths.mechanism-profile-conformance/1",
    implementation,
    implementationVersion: version,
    runtime,
    capabilities,
    results: Object.freeze(results),
    passed: results.every((result) => result.passed),
    claim: "test-results-only-not-security-certification",
  });
}

function reservation(key: string, byte: number, value: Uint8Array): AtomicReservationRecord {
  return Object.freeze({ key, commitment: new Uint8Array(32).fill(byte), value });
}

function transportOptions(maximumBytes: number): Readonly<{ maximumBytes: number; signal: AbortSignal }> {
  return Object.freeze({ maximumBytes, signal: new AbortController().signal });
}

function equalBytes(left: Uint8Array, right: Uint8Array): boolean {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

function bounded(value: string, name: string): string {
  if (typeof value !== "string" || value.length === 0 || value.length > 128 || /[\r\n]/u.test(value)) {
    throw new TypeError(`invalid conformance ${name}`);
  }
  return value;
}
